### PR TITLE
create envelope icon

### PR DESCRIPTION
**Pull request by erikdstock**
This is a first pass at creating an icon component from an SVG. It's essentially 

Icons are tricky because you often want to change their color. I solved this by taking the default heroicons.com solid icon svg and removing a few properties, then adding others that would allow it to inherit from the parent. It thus takes no incoming props and expects to be used inline with a container defining a `font-size` and `color`; Interested in others' thoughts on how we could do this, especially considering the use of tailwind or other libraries that define a color scheme. Perhaps we could allow overriding the `currentcolor` and width/height properties via a prop?

image

- uses [heroicons.com solid version](https://heroicons.com/)
- respects parent font-size and color properties
- some alignment with parent text is not perfect yet

### CODE REVIEW

**Comment by erikdstock** [comment]
I had this idea that i could have a folder like `icons/svgs` with every icon as an svg file in it, and I could dynamically import them all and iterate through, turning like `envelope-fill.svg` into `<dsa-icon-envelope-fill />` and so on and so forth. There's probably some kind of HOC that could accomplish that but i figured it wasn't necessary for now.

**Comment by nedredmond** [comment]
@erikdstock Interesting approach! I actually implemented an icon web component using a [FontAwesome icon](https://fontawesome.com/icons) kit-- it was one component that would load different icons based on a `Prop()` field, so it would look like:
`<DsaIcon name="envelope" style="solid" color="dsa-red" size="small">`
Each prop would use a TS string union type to define the range of acceptable inputs.

The idea about scanning the directory to generate individual components is a fun one-- one that I tried to do dynamically client-side with the font directory. Turns out you can't do that! Node.js `fs` isn't available and I couldn't find a working shim.

But maybe we could use that approach to dynamically generate the string literal union type and a mapper that would fetch the SVG from the provided name? Then it would look like this (and apologies if my syntax isn't perfect, we use Flow types at my current place of work :roll_eyes: )

```ts
// icon/types.ts
const iconNames = ["envelope", "user", "users"] as const;

export type IconName = typeof iconNames[number]
export const iconMap: { [name: IconName]: string } = {
  "envelope": ".../svg/envelope.svg",
  "user": ".../svg/user.svg",
  "etc": "...and/so.on"
};
```
...
```ts
// import types, and within component...
()Prop name: IconName;

// ...

render() {
  return <object type="image/svg+xml" data="{iconMap[this.name]}" />
}
```

That way, there's one icon component with a simple API to render different icons. We can manage identical JS and styles in one place without worrying about making sure our dozens of icon components are synced. What do you think?

**Comment by ysulyma** [comment]
I think we should try to stick with native web semantics where we can. For icons, why create a component instead of just using `<svg>` or `<object>`?

There are a few ways to adjust colors in SVG to match the surrounding text:
- the easiest is to include the SVGs inline and use `{fill,stroke}="currentColor"` in the SVG file. Bundlers will do this for you, e.g. [`@svgr/webpack`](https://www.npmjs.com/package/@svgr/webpack).
- if using an `<object>` to load the SVG, you can define CSS variables in the SVG file and then use Javascript to set them using `objectTag.contentDocument`. This only works if the SVG file is served from the same domain as the page embedding the `<object>` tag. This also causes a FOUC in loading the icon.
- can `fetch()` the SVG and modify the colors before inserting it in the document. This also suffers from FOUC.

**Comment by erikdstock** [comment]
> I think we should try to stick with native web semantics where we can. For icons, why create a component instead of just using <svg> or <object>? - @ysulyma 

We need to establish some common opinions and best practices for this repo, but my take here is that a WC system should primarily export a single package of web components and not just have a collection of files that consumers need to be able to interpret or load independently. I'll try to come up with a solution that avoids the flash of unstyled content.

**Comment by cewkrupa** [comment]
I like this approach! It feels simple enough to start.

With the caveat that I don't know much about WebComponents, I think there could be a good combo of @erikdstock's approach of

>I had this idea that i could have a folder like icons/svgs with every icon as an svg file in it, and I could dynamically import them all and iterate through, turning like envelope-fill.svg into <dsa-icon-envelope-fill /> and so on and so forth. 

and Ned's approach of 
> `<DsaIcon name="envelope" style="solid" color="dsa-red" size="small">`

for automatically generating basic icon components like `<dsa-icon-envelope-fill /` and then having a container component that you could pass props to. So something like:

```
<icon-container width=12 height=12 color="red">
  <dsa-icon-envelope-fill />
</icon-container>
```

I think this would have the benefits of
- providing a fairly straightforward example of usage
- allow consumers to override common styles and props as needed
- allow consumers to use just the svg icon with their own wrapper if they want more control over the appearance
- allow consumers to implement their own icons and use the same wrapper

I'm not sure what the downsides would be.

**Comment by nedredmond** [comment]
@cewkrupa Due to the encapsulation of the shadow dom, styles from the container would not get passed down to the icon.

**Comment by erikdstock** [comment]
Leaving myself some [breadcrumbs](https://www.sarasoueidan.com/blog/svg-style-inheritance-and-fousvg/) here on SVG Flash of unstyled content.

File: readme.md [hunk 1]
```diff
@@ -2,50 +2,74 @@
 
 [📖 dsausa.github.io/dsa-ui](https://dsausa.github.io/dsa-ui/)
 
-This is a design system built with [Stencil](https://stenciljs.com/) and [Storybook](https://storybook.js.org/) for Democratic Socialists of America sites and applications by the DSA National Tech Committee.
+This is a design system built with [Stencil](https://stenciljs.com/) and
```
Comment (nedredmond) [inline readme.md:5]: Why so many random line breaks and formatting changes here?

**Review by nedredmond — COMMENTED** [review]
A little confused by the readme changes which are hard to parse. Can you fill me in on what happened there? Everything else looks fine.

File: readme.md (hunk 1 above)
Comment (erikdstock) [reply readme.md:5]: ah, prettier. I can undo that if we dont like it. i find it very hard to get my config just right and i guess it's on by default now, but it causes an annoying huge diff when i touch an unformatted md file (format on focus lost). I can disable it for this repo and undo.

Comment (nedredmond) [reply readme.md:5]: I don't think we want to line-wrap at 80 characters if that's what's happening here. The prettier config is in `./wc` and ignores `*.md`. Is this a global setting for you?

Comment (erikdstock) [reply readme.md:5]: Alright, reverted. FWIW I don't think the prettier config will apply to directories above the folder it's in. Any aversion to changing the name to `README.md` (all caps)? I'll leave that off for now since it's unrelated and sometimes causes issues in non-case-sensitive filesystems.

**Comment by nedredmond** [comment]
Strange that this is failing to build in CI. I'm going to pull it down and take a look.

**Comment by nedredmond** [comment]
I played with this for 2 hours and I have converted. I'm fine with multiple inline SVGsx.

**Comment by nedredmond** [comment]
As far as the CI goes, still working on that because I have no issues locally and the main branch is passing.

**Comment by nedredmond** [comment]
For some reason the Node 19 pipeline wasn't downloading Chromium in the `pnpm build` step, which is odd because it did in prior runs.

**Comment by nedredmond** [comment]
I think we should remove the dynamic icon component for now since it doesn't work as expected and focus on importing icons that we know we need.

**Comment by erikdstock** [comment]
One final thought for the single svg is adding an [svg inline loader](https://v4.webpack.js.org/loaders/svg-inline-loader/) - which might allow dynamic loading (like `const SVGComponent = await import(props.kind)`) and then we could add styles dynamically from there.

**Comment by ysulyma** [comment]
@erikdstock give it a shot, but I don't believe it's possible to use a dynamic path with any SVG inline loaders, since they operate at bundle time.
